src/msg/async/frames_v2.h

```diff
@@ -804,9 +804,12 @@ struct MessageFrame : public Frame<MessageFrame,
     return f;
   }
 
-  inline const ceph_msg_header2 &header() {
+  inline const ceph_msg_header2 header() {
     auto& hdrbl = segments[SegmentIndex::Msg::HEADER];
-    return reinterpret_cast<const ceph_msg_header2&>(*hdrbl.c_str());
+    bufferlist::const_iterator it = hdrbl.cbegin();
+    ceph_msg_header2 header;
+    decode(header, it);
```

Comment:
but this also incurs memcpy(). see the implementation of `decode_raw()`. but yeah, it helps to avoid rebuilding the whole chunk of `segments[SegmentIndex::Msg::HEADER]`. if its size is much greater than that of `ceph_msg_header2`, it's a great performance improvement. otherwise, i am not quite sure.

Reply:
Both implementations incur a memcpy of sizeof(ceph_msg_header2).
The difference is that the current solution also adds a call to malloc(sizeof(ceph_msg_header2)).

Reply:
makes sense! in this case, since the guaranteed copy elision kicks in, the caller is responsible to allocate the memory chunk (on stack) and leave it to the callee to materialize it. this save us a malloc() call.